ports/esp32/machine_pwm.c

```diff
@@ -119,74 +146,177 @@ STATIC void configure_channel(machine_pwm_obj_t *self) {
     }
 }
 
-STATIC void set_freq(int newval, ledc_timer_config_t *timer) {
-    // If already set, do nothing
-    if (newval == timer->freq_hz) {
-        return;
-    }
+STATIC void set_duty_u16(machine_pwm_obj_t *self, int duty);
+STATIC void set_duty_u10(machine_pwm_obj_t *self, int duty);
+STATIC void set_duty_ns(machine_pwm_obj_t *self, int ns);
 
-    // Find the highest bit resolution for the requested frequency
-    if (newval <= 0) {
-        newval = 1;
-    }
-    unsigned int res = 0;
-    for (unsigned int i = LEDC_APB_CLK_HZ / newval; i > 1; i >>= 1) {
-        ++res;
-    }
-    if (res == 0) {
-        res = 1;
-    } else if (res > PWRES) {
-        // Limit resolution to PWRES to match units of our duty
-        res = PWRES;
-    }
+STATIC void set_freq(unsigned int freq, ledc_timer_config_t *timer, machine_pwm_obj_t *self) {
+    // Even if the timer frequency is already set,
+    // the set_duty_x() is required to reconfigure the channel duty anyway
+    if (freq != timer->freq_hz) {
+        DBG(3, "set_freq(%d)", freq)
 
-    // Configure the new resolution and frequency
-    timer->duty_resolution = res;
-    timer->freq_hz = newval;
+        // Find the highest bit resolution for the requested frequency
+        unsigned int i = LEDC_APB_CLK_HZ; // 80 MHz
+        if (freq < EMPIRIC_FREQ) {
+            i = LEDC_REF_CLK_HZ; // 1 MHz
+        }
 
-    // set freq
-    esp_err_t err = ledc_timer_config(timer);
-    if (err != ESP_OK) {
-        if (err == ESP_FAIL) {
-            PWM_DBG("timer timer->speed_mode %d, timer->timer_num %d, timer->clk_cfg %d, timer->freq_hz  %d, timer->duty_resolution %d)", timer->speed_mode, timer->timer_num, timer->clk_cfg, timer->freq_hz, timer->duty_resolution);
-            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("bad frequency %d"), newval);
-        } else {
-            check_esp_err(err);
+        #if 1
+        // original code
+        i /= freq;
+        #else
+        // See https://github.com/espressif/esp-idf/issues/7722
+        unsigned int divider = i / freq; // truncated
+        // int divider = (i + freq / 2) / freq; // rounded
+        if (divider == 0) {
+            divider = 1;
+        }
+        float f = (float)i / divider; // actual frequency
+        if (f <= 1.0) {
+            f = 1.0;
+        }
+        i = (unsigned int)roundf((float)i / f);
```

Comment:
please remove this unused code

Reply:
This code is not redundant.
This code will use after resolving ESP-IDF issue
[Make LEDC frequency more accurate (IDFGH-6037)](https://github.com/espressif/esp-idf/issues/7722)